### PR TITLE
TE-208 Light node launcher

### PR DESCRIPTION
**Pull request by adonagy**
This PR is still a WIP. Few TODOs are still left to be done.

Will look into the possibility to reuse the Environment struct from the light node.

### CODE REVIEW

File: .drone.yml [hunk 1]
```diff
@@ -95,6 +96,84 @@ trigger:
...
+        path: /tmp/sandbox
+    environment:
+      SODIUM_USE_PKG_CONFIG: 1
+      RUST_BACKTRACE: 1
+    commands:
+      - cp /artifacts/protocol-runner /tmp/sandbox/
+      - rust_libs=$(echo "`rustup show home`/toolchains/`rustup show active-toolchain | tr " " "\n" | head -1`/lib")
+      - export LD_LIBRARY_PATH="/artifacts/ffi:$rust_libs"
+      - echo "LD_LIBRARY_PATH - $LD_LIBRARY_PATH"
+      - /artifacts/launcher --launcher-rpc-port 3030 --light-node-path /artifacts/light-node --logl-level info
```
Comment (bkontur) [inline .drone.yml:133]: @adonagy 
--logl-level vs --log-level :)

File: .drone.yml [hunk 2]
```diff
@@ -118,6 +197,19 @@ environment:
   BOOTSTRAP2_SECRET: unencrypted:edsk39qAm1fiMjgmPkw1EgQYkMzkJezLNewd7PLNHTkr6w9XA2zdfo
 
 steps:
+- name: cleanup-node-data
+  image: alpine/git
+  user: root
+  volumes:
```
Comment (bkontur) [inline .drone.yml:203]: @adonagy 
here should be just:
```
 volumes:
    - name: sandbox-data
      path: /tmp/sandbox
```

File: .drone.yml [hunk 3]
```diff
@@ -95,6 +96,84 @@ trigger:
...
+    - name: sandbox-data
+      path: /tmp/sandbox
+  commands:
+    - sleep 5
+    - LAUNCHER_URL=$${TEZEDGE_LAUNCHER_RPC_CONTEXT_ROOT}
+    - resp=$(curl --request POST "$LAUNCHER_URL/start --data-raw '{"config_file":"./light_node/etc/tezedge/tezedge.config","identity_expected_pow":0,"disable_bootstrap_lookup":true,"network":"sandbox","peer_thresh_low":1,"peer_thresh_high":1,"sandbox_patch_context_json_file":"./light_node/etc/tezedge_sandbox/sandbox-patch-context.json","protocol_runner":"./target/release/protocol-runner"}' " | jq -S .)
+    - sleep 5
+  
+- name: stop-tezedge-node-via-rpc
```
Comment (bkontur) [inline .drone.yml:147]: @adonagy 
to verify startup, you can use:

```
- name: wait-for-tezedge-to-start
  image: simplestakingcom/tezos-node-bootstrap:latest
  commands:
    - tezos-node-bootstrap --bootstrap --level=0 --node_tezedge=$${TEZEDGE_NODE_RPC_CONTEXT_ROOT}
```

Comment (adonagy) [reply .drone.yml:147]: Yeah, good idea.

File: sandbox/src/configuration.rs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+use clap::{App, Arg};
+
+pub struct LauncherEnvironment {
+    pub light_node_path: PathBuf,
+    pub log_level: slog::Level,
+    pub sandbox_rpc_port: u16,
+}
+
+// TODO: borrowed from light_node/src/configuration; expose this macro
```
Comment (bkontur) [inline sandbox/src/configuration.rs:11]: @adonagy 
just remove TODO pls., if you dont plan to fix/change it

File: sandbox/src/filters.rs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+use warp::Filter;
+
+use crate::handlers::{start_node_with_config, stop_node};
+use crate::node_runner::LightNodeRunnerRef;
+
+pub fn sandbox(
+    log: Logger,
+    runner: LightNodeRunnerRef,
+) -> impl Filter<Extract = impl warp::Reply, Error = warp::Rejection> + Clone {
+    start(log.clone(), runner.clone()).or(stop(log.clone(), runner))
```
Comment (bkontur) [inline sandbox/src/filters.rs:11]: @adonagy 
I think, you dont need to clone second logger `stop(log.clone(), runner)` -> `stop(log, runner)`

File: sandbox/src/handlers.rs [hunk 6]
```diff
@@ -0,0 +1,54 @@
+use std::convert::Infallible;
+
+use slog::{error, info, Logger};
+use warp::http::StatusCode;
+
+use crate::node_runner::LightNodeRunnerRef;
+
+// TODO: discussion about what status codes to return on errors
```
Comment (bkontur) [inline sandbox/src/handlers.rs:8]: @adonagy 
lets discuss or just remove it :D

File: sandbox/src/node_runner.rs [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+/// Thread safe reference to a shared Runner
+pub type LightNodeRunnerRef = Arc<RwLock<LightNodeRunner>>;
+
+/// Struct that holds info about the running child process
+pub struct LightNodeRunner {
+    executable_path: PathBuf,
+    _name: String,
+    process: Option<Child>,
+    // TODO: anything else? Do we need a name? Maybe in the furture when launching multiple nodes with the sandbox
```
Comment (bkontur) [inline sandbox/src/node_runner.rs:44]: @adonagy 
I think we should support multiple nodes now

File: sandbox/src/filters.rs [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+fn json_body() -> impl Filter<Extract = (serde_json::Value,), Error = warp::Rejection> + Clone {
+    // When accepting a body, we want a JSON body
+    // (and to reject huge payloads)...
+    warp::body::content_length_limit(1024 * 16).and(warp::body::json())
+}
+
+fn with_log(
+    log: Logger,
+) -> impl Filter<Extract = (Logger,), Error = std::convert::Infallible> + Clone {
+    warp::any().map(move || log.clone())
```
Comment (bkontur) [inline sandbox/src/filters.rs:46]: @adonagy 
is this log.clone necessery?

File: sandbox/src/filters.rs [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+fn with_runner(
+    runner: LightNodeRunnerRef,
+) -> impl Filter<Extract = (LightNodeRunnerRef,), Error = std::convert::Infallible> + Clone {
+    warp::any().map(move || runner.clone())
```
Comment (bkontur) [inline sandbox/src/filters.rs:52]: @adonagy 
is this runner.clone necessery?

File: sandbox/src/node_runner.rs (hunk 7 above)
Comment (adonagy) [reply sandbox/src/node_runner.rs]: Let's merge this with just one node support for now. I created a task to add multinode support.

File: sandbox/src/filters.rs (hunk 8 above)
Comment (adonagy) [reply sandbox/src/filters.rs:46]: It is as we are turning that clone-able resource into a Filter (warp [docs](https://docs.rs/warp/0.2.4/warp/filters/any/fn.any.html))

File: sandbox/src/filters.rs (hunk 9 above)
Comment (adonagy) [reply sandbox/src/filters.rs:52]: Same as log.

File: sandbox/src/tezos_client_runner.rs [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+    }
+
+    pub fn init_client_data(&self) -> Result<(), TezosClientRunnerError> {
+        let base_dir = if let Some(path) = self.base_dir_path.to_str() {
+            println!("PATH: {}", path);
+            path
+        } else {
+            return Err(TezosClientRunnerError::PathError);
+        };
+        println!("Base dir: {}", base_dir);
```
Comment (bkontur) [inline sandbox/src/tezos_client_runner.rs:109]: @adonagy 
Could you pls. change println to slog?

Comment (adonagy) [reply sandbox/src/tezos_client_runner.rs:109]: Just a stray printl, will remove

**Pull request merged**